### PR TITLE
cool easter egg ig

### PR DESCRIPTION
**Pull request by UntouchedOdin0**

### CODE REVIEW

File: src/main/kotlin/net/developerden/devdenbot/listener/DiscordTokenMessageListener.kt [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+import javax.inject.Inject
+
+class DiscordTokenMessageListener @Inject constructor(
+    private val ddbConfig: DDBConfig,
+    ) : EventListener {
+
+    companion object {
+        private val log by log()
+
+        internal val tokenRegex = Regex("[MN][A-Za-z\\d]{23}\\.[\\w-]{6}\\.[\\w-]{27}\n")
```
Comment (knightzmc) [inline src/main/kotlin/net/developerden/devdenbot/listener/DiscordTokenMessageListener.kt:20]: image

This regex doesn't seem to actually work

File: src/main/kotlin/net/developerden/devdenbot/listener/DiscordTokenMessageListener.kt [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+    }
+
+    private suspend fun onGuildMessageReceived(event: GuildMessageReceivedEvent) {
+        if (event.author.isBot) {
+            return
+        }
+        if (event.channel.idLong == BOT_COMMANDS_CHANNEL_ID) {
```
Comment (knightzmc) [inline src/main/kotlin/net/developerden/devdenbot/listener/DiscordTokenMessageListener.kt:27]: There's no need to do this, the `ddkeep` check, or the `/run` check - copy pasting is fine but make sure to trim out the unnecessary bits

File: src/main/kotlin/net/developerden/devdenbot/listener/DiscordTokenMessageListener.kt [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+        if (rawText.startsWith("${ddbConfig.prefix}keep")) {
+            return
+        }
+        if (rawText.startsWith("/run")){
+            return
+        }
+        if (!rawText.contains(tokenRegex)) {
+            return
+        }
+        if (rawText.contains(tokenRegex)) {
```
Comment (knightzmc) [inline src/main/kotlin/net/developerden/devdenbot/listener/DiscordTokenMessageListener.kt:40]: this will always be true because you've already checked `!rawText.contains(tokenRegex)`

File: src/main/kotlin/net/developerden/devdenbot/listener/DiscordTokenMessageListener.kt [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+            event.message.delete()
```
Comment (knightzmc) [inline src/main/kotlin/net/developerden/devdenbot/listener/DiscordTokenMessageListener.kt:41]: you need to .await() or .queue() the delete

File: src/main/kotlin/net/developerden/devdenbot/listener/DiscordTokenMessageListener.kt [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+            event.channel.sendMessage("The previous message has been deleted due to").await()
```
Comment (knightzmc) [inline src/main/kotlin/net/developerden/devdenbot/listener/DiscordTokenMessageListener.kt:42]: why 2 separate messages?

File: src/main/kotlin/net/developerden/devdenbot/listener/JoinLeaveListener.kt [hunk 6]
```diff
@@ -15,8 +17,18 @@ import net.dv8tion.jda.api.events.guild.member.GuildMemberRemoveEvent
 class JoinLeaveListener : EventListener {
     private suspend fun onGuildMemberJoin(event: GuildMemberJoinEvent) {
         val channel = event.jda.getTextChannelById(WELCOME_CHANNEL_ID)
+        val memberCount = event.guild.memberCount
+        val member = event.member
+        val guild: Guild = event.guild
+        val role: Role? = guild.getRoleById(666) // TODO ALEX please put the role idea for 666 :)
```
Comment (knightzmc) [inline src/main/kotlin/net/developerden/devdenbot/listener/JoinLeaveListener.kt:23]: i'm kinda relucant to add easter eggs like this just because whilst they're cool, they'll just lead to more people wanting funny gimmick roles. the meme roles were a similar idea, but in the end they just look messy. 
also, this could lead to multiple people getting the role, which probably isn't what we want

File: src/main/kotlin/net/developerden/devdenbot/pasting/CodeBlockMessageListener.kt [hunk 7]
```diff
@@ -46,6 +45,7 @@ class CodeBlockMessageListener @Inject constructor(
         if (!rawText.contains(codeBlock)) {
             return
         }
+
```
Comment (knightzmc) [inline src/main/kotlin/net/developerden/devdenbot/pasting/CodeBlockMessageListener.kt:48]: try and avoid unnecessary changes in a PR

File: src/main/kotlin/net/developerden/devdenbot/listener/BumpNotificationListener.kt [hunk 8]
```diff
@@ -64,6 +64,7 @@ class BumpNotificationListener : EventListener {
         event.channel.sendMessage(
             "${bumpNotificationRole.asMention}, the server is ready to be bumped! **!d bump**"
         ).await()
+        return
```
Comment (knightzmc) [inline src/main/kotlin/net/developerden/devdenbot/listener/BumpNotificationListener.kt:67]: As mentioned in the other PR, this doesn't actually do anything and won't fix the issue

File: src/main/kotlin/net/developerden/devdenbot/DevDen.kt [hunk 9]
```diff
@@ -112,7 +112,7 @@ class DevDen {
     }
 
     private fun startTasks(jda: JDA) {
-        log.info { "starting tasks..." }
+        log.info { "Starting tasks..." }
```
Comment (knightzmc) [inline src/main/kotlin/net/developerden/devdenbot/DevDen.kt:115]: try and avoid unnecessary changes in a PR

File: src/main/kotlin/net/developerden/devdenbot/commands/info/StackOverflowCommand.kt [hunk 10]
```diff
@@ -0,0 +1,32 @@
...
+ */
+
+@Used
+class StackOverflowCommand constructor(
+    override val ddbConfig: DDBConfig,
+) : DevDenSlashCommand(
+    name = "stackoverflow",
+    description = "A helpful way to find answers using stackoverflow"
+), HasConfig {
+    override suspend fun load(action: CommandCreateAction) = Unit
```
Comment (knightzmc) [inline src/main/kotlin/net/developerden/devdenbot/commands/info/StackOverflowCommand.kt:23]: I'm happy to add this, but I don't think it will work. 
You need to register the command arguments inside `load`

File: src/main/kotlin/net/developerden/devdenbot/commands/info/StackOverflowCommand.kt [hunk 11]
```diff
@@ -0,0 +1,32 @@
...
+
+    override suspend fun SlashCommandEvent.execute() {
+        val question = getOption("question")?.asString ?: error("Invalid question")
+        val response = URL(question.replace("https://api.stackexchange.com/2.3/search/advanced?order=desc&sort=activity&q=$question",
```
Comment (knightzmc) [inline src/main/kotlin/net/developerden/devdenbot/commands/info/StackOverflowCommand.kt:27]: Similarly, I don't think this will work. You're replacing the URL inside `question`, which is just gonna be a plain string, so the replacement will do nothing. I think you can just change this to `URL("https://api.stackexchange.com/2.3/search/advanced?order=desc&sort=activity&q=$question")`

**Review by knightzmc — CHANGES_REQUESTED** [review]
Summary:

- I'm happy to add the StackOverflow command because it might come in handy, but make sure it works
- Don't really want to add more gimmick roles, so I'd ask that you remove the 666 stuff (perhaps we could do it with a nickname instead)
- DiscordTokenMessageListener is mostly good, but the regex and the logic need a look over
- For sake of consistency, can you remove all of the changes that aren't directly related to the token blocking, or the stackoverflow command

Thanks

File: src/main/kotlin/net/developerden/devdenbot/pasting/CodeBlockMessageListener.kt [hunk 12]
```diff
@@ -46,6 +46,12 @@ class CodeBlockMessageListener @Inject constructor(
         if (!rawText.contains(codeBlock)) {
             return
         }
+        if (rawText.contains(tokenRegex)) {
+            event.message.delete()
+            event.channel.sendMessage("The previous message has been deleted due to").await()
+            event.channel.sendMessage("it containing a discord bot token!").await()
```
Comment (DrZoddiak) [inline src/main/kotlin/net/developerden/devdenbot/pasting/CodeBlockMessageListener.kt]: I think we should make it clear that the bot token should also be changed.
Once it has been sent, it should be considered compromised.

File: src/main/kotlin/net/developerden/devdenbot/listener/DiscordTokenMessageListener.kt (hunk 5 above)
Comment (UntouchedOdin0) [reply src/main/kotlin/net/developerden/devdenbot/listener/DiscordTokenMessageListener.kt:42]: I thought it was too long?

**Comment by knightzmc** [comment]
Closing due to inactivity
